src/libstd/sys/unix/fs.rs

```diff
@@ -869,18 +870,35 @@ pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
         )
     }
 
-    if !from.is_file() {
-        return Err(Error::new(ErrorKind::InvalidInput,
-                              "the source path is not an existing regular file"))
-    }
-
     let mut reader = File::open(from)?;
-    let mut writer = File::create(to)?;
-    let (perm, len) = {
+
+    let (mode, len) = {
         let metadata = reader.metadata()?;
-        (metadata.permissions(), metadata.size())
+        if !metadata.is_file() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "the source path is not an existing regular file",
+            ));
+        }
+        (metadata.permissions().mode(), metadata.len())
     };
 
+    let mut writer = OpenOptions::new()
+        // prevent world readable/writeable file in case of empty umask
```

Comment:
I find this a bit cryptic, but could this be expanded to explain what's going on here?